File: mcp-server/rynix_mcp/stack_detect.py
"""Detect application stack from repository manifests."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any


def _read_json(path: Path) -> dict[str, Any] | None:
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None


def detect_stack(repo_path: str | Path) -> list[dict[str, Any]]:
    repo = Path(repo_path).resolve()
    detected: list[dict[str, Any]] = []

    pkg = repo / "package.json"
    if pkg.is_file():
        data = _read_json(pkg) or {}
        deps = {**data.get("dependencies", {}), **data.get("devDependencies", {})}
        if "next" in deps:
            detected.append(
                {"name": "nextjs", "confidence": "high", "signals": ["package.json:next"]}
            )
        elif "react" in deps:
            detected.append(
                {"name": "react", "confidence": "high", "signals": ["package.json:react"]}
            )
        if "@nestjs/core" in deps:
            detected.append(
                {"name": "nestjs", "confidence": "high", "signals": ["package.json:@nestjs/core"]}
            )
        if "express" in deps:
            detected.append(
                {"name": "express", "confidence": "high", "signals": ["package.json:express"]}
            )
        if "fastify" in deps:
            detected.append(
                {"name": "fastify", "confidence": "high", "signals": ["package.json:fastify"]}
            )
        if "vue" in deps:
            detected.append(
                {"name": "vue", "confidence": "medium", "signals": ["package.json:vue"]}
            )
        if "nuxt" in deps or "nuxt3" in deps:
            detected.append(
                {"name": "nuxt", "confidence": "high", "signals": ["package.json:nuxt"]}
            )
        if "@angular/core" in deps:
            detected.append(
                {"name": "angular", "confidence": "high", "signals": ["package.json:@angular/core"]}
            )
        if "@sveltejs/kit" in deps:
            detected.append(
                {
                    "name": "sveltekit",
                    "confidence": "high",
                    "signals": ["package.json:@sveltejs/kit"],
                }
            )

    if (repo / "go.mod").is_file():
        text = (repo / "go.mod").read_text(encoding="utf-8", errors="ignore")
        if "gin-gonic/gin" in text:
            detected.append({"name": "gin", "confidence": "high", "signals": ["go.mod:gin"]})
        elif "labstack/echo" in text:
            detected.append({"name": "echo", "confidence": "high", "signals": ["go.mod:echo"]})
        elif "gofiber/fiber" in text:
            detected.append({"name": "fiber", "confidence": "high", "signals": ["go.mod:fiber"]})
        elif "go-chi/chi" in text:
            detected.append({"name": "chi", "confidence": "high", "signals": ["go.mod:chi"]})

    if (repo / "pyproject.toml").is_file() or (repo / "requirements.txt").is_file():
        py_text = ""
        if (repo / "pyproject.toml").is_file():
            py_text += (repo / "pyproject.toml").read_text(encoding="utf-8", errors="ignore")
        if (repo / "requirements.txt").is_file():
            py_text += (repo / "requirements.txt").read_text(encoding="utf-8", errors="ignore")
        if "fastapi" in py_text.lower():
            detected.append(
                {"name": "fastapi", "confidence": "high", "signals": ["python:fastapi"]}
            )
        if "django" in py_text.lower():
            detected.append({"name": "django", "confidence": "high", "signals": ["python:django"]})
        if "flask" in py_text.lower():
            detected.append({"name": "flask", "confidence": "medium", "signals": ["python:flask"]})

    if (repo / "composer.json").is_file():
        data = _read_json(repo / "composer.json") or {}
        req = data.get("require", {})
        if "laravel/framework" in req:
            detected.append(
                {"name": "laravel", "confidence": "high", "signals": ["composer.json:laravel"]}
            )
        if "symfony/framework-bundle" in req:
            detected.append(
                {"name": "symfony", "confidence": "high", "signals": ["composer.json:symfony"]}
            )

    if (repo / "Gemfile").is_file():
        text = (repo / "Gemfile").read_text(encoding="utf-8", errors="ignore")
        if "rails" in text:
            detected.append({"name": "rails", "confidence": "high", "signals": ["Gemfile:rails"]})
        if "sinatra" in text:
            detected.append(
                {"name": "sinatra", "confidence": "medium", "signals": ["Gemfile:sinatra"]}
            )

    if (repo / "pom.xml").is_file() or list(repo.glob("**/build.gradle*")):
        detected.append(
            {"name": "spring", "confidence": "medium", "signals": ["jvm:spring-or-gradle"]}
        )

    if list(repo.glob("**/*.csproj")):
        detected.append({"name": "aspnet", "confidence": "medium", "signals": ["csproj"]})

    if (repo / "Cargo.toml").is_file():
        text = (repo / "Cargo.toml").read_text(encoding="utf-8", errors="ignore")
        if "axum" in text:
            detected.append({"name": "axum", "confidence": "high", "signals": ["Cargo.toml:axum"]})
        if "actix-web" in text:
            detected.append(
                {"name": "actix", "confidence": "high", "signals": ["Cargo.toml:actix"]}
            )

    if not detected:
        detected.append({"name": "generic", "confidence": "low", "signals": ["no_manifest_match"]})

    return detected

File: rynix-core/src/stack_detect.rs
use std::path::{Path, PathBuf};

/// Lightweight stack detection from repository manifests (mirrors stack_detect.py).
pub fn detect_stacks(repo: &Path) -> Vec<String> {
    let mut stacks: Vec<String> = Vec::new();

    let pkg = repo.join("package.json");
    if pkg.is_file() {
        if let Ok(text) = std::fs::read_to_string(&pkg) {
            if text.contains("\"next\"") {
                stacks.push("nextjs".into());
            } else if text.contains("\"react\"") {
                stacks.push("react".into());
            }
            if text.contains("@nestjs/core") {
                stacks.push("nestjs".into());
            }
            if text.contains("\"express\"") {
                stacks.push("express".into());
            }
            if text.contains("\"fastify\"") {
                stacks.push("fastify".into());
            }
            if text.contains("\"vue\"") {
                stacks.push("vue".into());
            }
            if text.contains("\"nuxt\"") || text.contains("\"nuxt3\"") {
                stacks.push("nuxt".into());
            }
            if text.contains("@angular/core") {
                stacks.push("angular".into());
            }
            if text.contains("@sveltejs/kit") {
                stacks.push("sveltekit".into());
            }
            if text.contains("@trpc/server") {
                stacks.push("trpc".into());
            }
            if text.contains("graphql") {
                stacks.push("graphql".into());
            }
            if text.contains("\"koa\"") {
                stacks.push("koa".into());
            }
            if text.contains("hono") {
                stacks.push("hono".into());
            }
            if text.contains("@adonisjs/core") {
                stacks.push("adonisjs".into());
            }
        }
    }

    let go_mod = repo.join("go.mod");
    if go_mod.is_file() {
        if let Ok(text) = std::fs::read_to_string(&go_mod) {
            if text.contains("gin-gonic/gin") {
                stacks.push("gin".into());
            } else if text.contains("labstack/echo") {
                stacks.push("echo".into());
            } else if text.contains("gofiber/fiber") {
                stacks.push("fiber".into());
            } else if text.contains("go-chi/chi") {
                stacks.push("chi".into());
            }
        }
    }

    let pyproject = repo.join("pyproject.toml");
    let requirements = repo.join("requirements.txt");
    let mut py_text = String::new();
    if pyproject.is_file() {
        if let Ok(t) = std::fs::read_to_string(&pyproject) {
            py_text.push_str(&t);
        }
    }
    if requirements.is_file() {
        if let Ok(t) = std::fs::read_to_string(&requirements) {
            py_text.push_str(&t);
        }
    }
    let py_lower = py_text.to_lowercase();
    if py_lower.contains("fastapi") {
        stacks.push("fastapi".into());
    }
    if py_lower.contains("django") {
        stacks.push("django".into());
    }
    if py_lower.contains("flask") {
        stacks.push("flask".into());
    }

    let composer = repo.join("composer.json");
    if composer.is_file() {
        if let Ok(text) = std::fs::read_to_string(&composer) {
            if text.contains("laravel/framework") {
                stacks.push("laravel".into());
            }
            if text.contains("symfony/framework-bundle") {
                stacks.push("symfony".into());
            }
        }
    }

    let gemfile = repo.join("Gemfile");
    if gemfile.is_file() {
        if let Ok(text) = std::fs::read_to_string(&gemfile) {
            if text.contains("rails") {
                stacks.push("rails".into());
            }
            if text.contains("sinatra") {
                stacks.push("sinatra".into());
            }
        }
    }

    if repo.join("pom.xml").is_file() || !glob_paths(repo, "**/build.gradle*").is_empty() {
        stacks.push("spring".into());
    }

    if !glob_paths(repo, "**/*.csproj").is_empty() {
        stacks.push("aspnet".into());
    }

    let cargo = repo.join("Cargo.toml");
    if cargo.is_file() {
        if let Ok(text) = std::fs::read_to_string(&cargo) {
            if text.contains("axum") {
                stacks.push("axum".into());
            }
            if text.contains("actix-web") {
                stacks.push("actix".into());
            }
        }
    }

    if repo.join("mix.exs").is_file() {
        stacks.push("phoenix".into());
    }

    if repo.join("openapi.yaml").is_file()
        || repo.join("openapi.json").is_file()
        || repo.join("swagger.json").is_file()
    {
        stacks.push("openapi".into());
    }

    stacks.sort();
    stacks.dedup();
    stacks
}

fn glob_paths(repo: &Path, _pattern: &str) -> Vec<PathBuf> {
    let mut out = Vec::new();
    if _pattern.contains("csproj") {
        for entry in walkdir::WalkDir::new(repo)
            .into_iter()
            .filter_map(|e| e.ok())
            .filter(|e| e.file_type().is_file())
        {
            if entry.path().extension().is_some_and(|x| x == "csproj") {
                out.push(entry.path().to_path_buf());
            }
        }
    }
    if _pattern.contains("gradle") {
        for entry in walkdir::WalkDir::new(repo)
            .into_iter()
            .filter_map(|e| e.ok())
            .filter(|e| e.file_type().is_file())
        {
            let name = entry.file_name().to_string_lossy();
            if name.starts_with("build.gradle") {
                out.push(entry.path().to_path_buf());
            }
        }
    }
    out
}
